temporal-sdk/src/main/java/io/temporal/internal/client/IteratorWithOffset.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.internal.client;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+class IteratorWithOffset<T> implements Iterator<T> {
+  private Iterator<T> iterator;
+  private Iterator<T> headIterator;
+  private boolean overflow = false;
+
+  public IteratorWithOffset(List<T> list, int startingIndex) {
+    // startingIndex theoretically can be out of bound at the moment we got here already, so
+    // we need to write implementation that tolerates that
+    try {
+      headIterator = list.subList(0, startingIndex).iterator();
+    } catch (IndexOutOfBoundsException e) {
+      headIterator = list.iterator();
+    }
+    // Initializations of headIterator and iterator are not atomic, safe.
+    // It can lead to an item to be skipped or subbed in some rare edge case
+    // if the collection is modified in just the right time. Considering it not a big deal.
```

Comment:
Agreed not a big deal, but if you can trust the list has random access (which it seems like you can), just store the list and the index as fields and increment/wrap manually instead of two iterators. Can throw no such element when the index reaches start a second time (keep bool if needed to track that). I think it's easier to understand code.

Granted, I'd be tempted at the known cost of performance for every start workflow, to just copy to array list and shuffle so you can have real random. With this implementation, if two worker factories that can accept the same workflow are next to each other in a list of 20, you'll have to land right on the starting index of the second one or the first one will be prioritized 19 times out of 20 (assuming uniform random).

Reply:
> Agreed not a big deal, but if you can trust the list has random access (which it seems like you can), just store the list and the index as fields and increment/wrap manually instead of two iterators.

Not much simpler, because we will have to handle all the out-of-bound exceptions, handle overflow and implement hasNext ourselves. I don't think it actually will give a simpler implementation.

> With this implementation, if two worker factories that can accept the same workflow are next to each other in a list of 20, you'll have to land right on the starting index of the second one or the first one will be prioritized 19 times out of 20 (assuming uniform random).

Honestly, in vast majority of cases, the first selected factory will be the last. Because we deregister factories when we shut them down. So pretty much the only way the described scenario is possible is when there are many WorkerFactories and most of them are paused or out of slots. And the "good" factories need to be next to each other. Very-very edgy.

But maybe I will just implement it with a array-copy and Collections.shuffle honestly for simplicity of reading and maintaining, because rare users should have more than one WorkflowFactory on one client anyway.

Reply:
We do need to randomize. Otherwise imagine a user that currently stats two WorkerFactories on the same client. For some strange reason, but we totally allow it. only one of these workers will get tasks from the client without randomization. So I think the fact of randomization is important, but for this situation only pretty much.